pulp_rpm/tests/functional/conftest.py

```diff
@@ -109,6 +111,64 @@ def signed_artifact(http_get, artifacts_api_client, gen_object_with_cleanup, tmp
     return gen_object_with_cleanup(artifacts_api_client, temp_file).to_dict()
 
 
+@pytest.fixture
+def rpm_artifact_factory(
+    http_get, artifacts_api_client, gen_object_with_cleanup, tmp_path, pulp_domain_enabled
+):
+    """Return an artifact created from uploading an RPM file."""
+
+    def _rpm_artifact_factory(url=RPM_SIGNED_URL, pulp_domain=None):
+        temp_file = tmp_path / str(uuid.uuid4())
+        temp_file.write_bytes(http_get(url))
+        kwargs = {}
+        if pulp_domain:
+            if not pulp_domain_enabled:
+                raise RuntimeError("Server does not have domains enabled.")
+            kwargs["pulp_domain"] = pulp_domain
+        return gen_object_with_cleanup(artifacts_api_client, temp_file, **kwargs).to_dict()
+
+    return _rpm_artifact_factory
+
+
+@pytest.fixture
+def rpm_package_factory(
+    http_get,
+    rpm_package_api,
+    add_to_cleanup,
+    monitor_task,
+    pulp_domain_enabled,
+):
+    """Return a Package created from uploading an RPM file."""
+
+    def _rpm_package_factory(url=RPM_SIGNED_URL, pulp_domain=None):
+        kwargs = {}
+        if pulp_domain:
+            if not pulp_domain_enabled:
+                raise RuntimeError("Server does not have domains enabled.")
+            kwargs["pulp_domain"] = pulp_domain
+
+        with NamedTemporaryFile() as file_to_upload:
+            file_to_upload.write(http_get(url))
+            upload_attrs = {"file": file_to_upload.name}
+            upload = rpm_package_api.create(**upload_attrs, **kwargs)
+
+        try:
+            upload_task_data = monitor_task(upload.task)
+        except PulpTaskError:
+            pass
+
+        for created_resource in upload_task_data.created_resources:
+            try:
+                new_obj = rpm_package_api.read(created_resource)
+            except Exception:
+                pass  # This isn't the right created_resource for this api_client
+            else:
+                add_to_cleanup(rpm_package_api, new_obj.pulp_href)
+                return new_obj.to_dict()
```

Comment:
If the upload fails, this will still fail once it gets to the `for` loop. I think it is fine to let it fail, no need to catch the exception.  Also, can we remove `to_dict()`.

Reply:
Lots of refactoring happened, this went away along the path.